great_expectations/datasource/fluent/sql_datasource.py

```diff
@@ -580,76 +580,87 @@ def _fully_specified_batch_requests(self, batch_request: BatchRequest) -> List[B
 
-        batch_list: List[Batch] = []
         if batch_request.partitioner:
             sql_partitioner = self.get_partitioner_implementation(batch_request.partitioner)
         else:
             sql_partitioner = None
+
         batch_spec_kwargs: dict[str, str | dict | None]
-        for request in self._fully_specified_batch_requests(batch_request):
-            batch_metadata: BatchMetadata = self._get_batch_metadata_from_batch_request(
-                batch_request=request
+        requests = self._fully_specified_batch_requests(batch_request)
+        unsorted_metadata_dicts = [self._get_batch_metadata_from_batch_request(r) for r in requests]
+        if sql_partitioner:
+            sorted_metadata_dicts = self.sort_batch_identifiers_list(
+                unsorted_metadata_dicts, sql_partitioner
             )
-            batch_spec_kwargs = self._create_batch_spec_kwargs()
-            if sql_partitioner:
-                batch_spec_kwargs["partitioner_method"] = sql_partitioner.method_name
-                batch_spec_kwargs["partitioner_kwargs"] = (
-                    sql_partitioner.partitioner_method_kwargs()
-                )
-                # mypy infers that batch_spec_kwargs["batch_identifiers"] is a collection, but
-                # it is hardcoded to a dict above, so we cast it here.
-                cast(Dict, batch_spec_kwargs["batch_identifiers"]).update(
-                    sql_partitioner.batch_parameters_to_batch_spec_kwarg_identifiers(
-                        request.options
-                    )
-                )
-            # Creating the batch_spec is our hook into the execution engine.
-            batch_spec = self._create_batch_spec(batch_spec_kwargs)
-            execution_engine: SqlAlchemyExecutionEngine = self.datasource.get_execution_engine()
-            data, markers = execution_engine.get_batch_data_and_markers(batch_spec=batch_spec)
-
-            # batch_definition (along with batch_spec and markers) is only here to satisfy a
-            # legacy constraint when computing usage statistics in a validator. We hope to remove
-            # it in the future.
-            # imports are done inline to prevent a circular dependency with core/batch.py
-            from great_expectations.core import IDDict
-            from great_expectations.core.batch import LegacyBatchDefinition
-
-            batch_definition = LegacyBatchDefinition(
-                datasource_name=self.datasource.name,
-                data_connector_name=_DATA_CONNECTOR_NAME,
-                data_asset_name=self.name,
-                batch_identifiers=IDDict(batch_spec["batch_identifiers"]),
-                batch_spec_passthrough=None,
+        else:
+            sorted_metadata_dicts = unsorted_metadata_dicts
+        sorted_metadata_dicts = sorted_metadata_dicts[batch_request.batch_slice]
+        batch_metadata = sorted_metadata_dicts[-1]
+        request_index = unsorted_metadata_dicts.index(batch_metadata)
```

Comment:
Can you add a comment that requests and unsorted_metadata_dicts or in the same order which is why we can get the correct request_index from this lookup.